__tests__/applyStyle.test.js

```diff
@@ -0,0 +1,237 @@
+/*eslint no-console: "off"*/
+import 'isomorphic-fetch';
+import nock from 'nock';
+import should from 'should/as-function';
+
+import VectorLayer from 'ol/layer/Vector';
+import VectorTileLayer from 'ol/layer/VectorTile';
+import ImageLayer from 'ol/layer/Image';
+import VectorTileSource from 'ol/source/VectorTile';
+import {createXYZ} from 'ol/tilegrid';
+
+import glStyle from './fixtures/osm-liberty/style.json';
+import glStyleLoresSprite from './fixtures/osm-liberty/osm-liberty.json';
+import glStyleHiresSprite from './fixtures/osm-liberty/osm-liberty@2x.json';
+import styleInvalidVersion from './fixtures/style-invalid-version.json';
+import styleEmptySprite from './fixtures/style-empty-sprite.json';
+import styleMissingSprite from './fixtures/style-missing-sprite.json';
+import styleInvalidSpriteURL from './fixtures/style-invalid-sprite-url.json';
+
+import {applyStyle} from '../index.js';
+
+
+// to verify that no external http calls are mistakenly made
+// protects against inadvertent typos in nock creation as well
+nock.disableNetConnect();
+
+
+describe('applyStyle style argument validation', function() {
+  const source = 'vector';
+  const layer = new VectorLayer();
+  const path = null;
+  const resolutions = null;
+
+  beforeEach(function() {
+    nock('https://rawgit.com')
+      .get('/maputnik/osm-liberty/gh-pages/sprites/osm-liberty.json')
+      .reply(200, glStyleLoresSprite)
+      .get('/maputnik/osm-liberty/gh-pages/sprites/osm-liberty.png')
+      .reply(200, '')
+      .get('/maputnik/osm-liberty/gh-pages/sprites/osm-liberty@2x.json')
+      .reply(200, glStyleHiresSprite)
+      .get('/maputnik/osm-liberty/gh-pages/sprites/osm-liberty@2x.png')
+      .reply(200, {});
+  });
+
+  afterEach(function() {
+    nock.cleanAll();
+  });
+
+  test('should handle valid style as JSON', function(done) {
+    applyStyle(layer, glStyle, source, path, resolutions).then(done).catch(done);
+  });
+
+  test('should handle valid style as JSON string', function(done) {
+    applyStyle(layer, JSON.stringify(glStyle), source, path, resolutions).then(done).catch(done);
+  });
+
+  test('should reject invalid style version', function(done) {
+    applyStyle(layer, styleInvalidVersion, source, path, resolutions).then(function() {
+      done(new Error('invalid style version promise should reject'));
+    }).catch(function(err) {
+      done();
+    });
+  });
+
+  test('should reject invalid ol layer type', function(done) {
+    const layer = new ImageLayer();
+    applyStyle(layer, glStyle, source, path, resolutions).then(function() {
+      done(new Error('invalid ol layer type promise should reject'));
+    }).catch(function(err) {
+      done();
+    });
+  });
+
+  // TODO: still need to figure out best way to validate sources
```

Comment:
Does this test actually fail? It would be nice to have it in the test suite.

Reply:
The main problem is I don't quite understand the codebase yet.

When `applyStyle()` is called from `apply()`, source ends up being an array with the mapbox id of every layer, so I think what would have to happen is each layer would have to be dereferenced to get it's source type and check it.

I'm hoping a better solution will present itself as I become more familiar with the codebase.  I have a vague idea that the actual source names are being accumulated somewhere and maybe that could be passed instead?

Can I put it off to a future PR, or do I need to dig into it now?  It should probably say "right way", not "best way".  :-)

Reply:
`apply()` optimizes the gl `layer` to ol `Source` mapping, i.e. it avoids creating multiple vector sources if the gl `layer`s use the same gl `source-layer`. So it can also call `applyStyle` several times with different subsets of the gl layers.

So the test you commented out here makes sense. To make it pass, you'd have to `reject` at the end of the "monster for-loop" as you call it, when `mapboxLayers` has a length of `0`.

Reply:
Let's handle this in a separate pull request.

Reply:
Addressed with #109.